src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,166 @@
+type Props = {|
+  i18n: I18nType,
+  permissions: Array<string>,
+|};
+
+type HostPermissionMessageType = 'domain' | 'site';
+type PermissionMessageType = HostPermissionMessageType | 'allUrls';
+
+type GetPermissionStringParams = {|
+  messageType: PermissionMessageType,
+  param?: string | number,
+  tooMany?: boolean,
+|};
+
+type GenerateHostPermissionsParams = {|
+  permissions: Array<string>,
+  // eslint-disable-next-line react/no-unused-prop-types
+  messageType: HostPermissionMessageType,
+|};
+
+export class HostPermissionsBase extends React.Component<Props> {
+  constructor(props: Props) {
+    super(props);
+    this.i18n = props.i18n;
+  }
+
+  getPermissionString({ messageType, param, tooMany = false }: GetPermissionStringParams): string {
+    // These should be kept in sync with Firefox's strings for webextention
+    // host permissions which can be found in
+    // https://hg.mozilla.org/mozilla-central/raw-file/tip/browser/locales/en-US/chrome/browser/browser.properties
+    switch (messageType) {
+      case 'allUrls':
+        return this.i18n.gettext('Access your data for all websites');
+      case 'domain':
+        if (tooMany) {
+          return this.i18n.sprintf(this.i18n.ngettext(
+            'Access your data in %(param)s other domain',
+            'Access your data in %(param)s other domains',
+            param), { param: this.i18n.formatNumber(param) }
+          );
+        }
+        return this.i18n.sprintf(
+          this.i18n.gettext('Access your data for sites in the %(param)s domain'),
+          { param }
+        );
+      case 'site':
+        if (tooMany) {
+          return this.i18n.sprintf(this.i18n.ngettext(
+            'Access your data on %(param)s other site',
+            'Access your data on %(param)s other sites',
+            param), { param: this.i18n.formatNumber(param) }
+          );
+        }
+        return this.i18n.sprintf(
+          this.i18n.gettext('Access your data for %(param)s'),
+          { param }
+        );
+      default:
+        return '';
+    }
+  }
+
+  i18n: I18nType;
+
+  // Generates Permission components for a list of host permissions. If we have 4 or
+  // fewer, display them all, otherwise display the first 3 followed by an item
+  // that says "...plus N others".
+  generateHostPermissions({
+    permissions, messageType,
+  }: GenerateHostPermissionsParams): Array<React.Element<typeof Permission>> {
+    const hostPermissions = [];
+    for (const item of permissions.slice(0, 4)) {
+      // Add individual Permission components for the first 4 host permissions.
+      hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.getPermissionString({ messageType, param: item })}
+          key={item}
+        />
+      );
+    }
+    if (permissions.length > 4) {
+      // Replace the final individual permission with a "too many" permission.
+      hostPermissions[3] = (<Permission
```

Comment:
Surprised eslint didn't catch this, but on multi-line components we'll start one after the `(`, so this should be:

```js
hostPermissions[3] = (
  <Permission
    type="hostPermission"
    [...etc.]
  />
);
```